### PR TITLE
Meta: Use cmake functions to uncompres zip/tarballs

### PR DESCRIPTION
**Pull request by elcuco**
Try to use CMake for extracting zip/tar.gz instead of forking external tools. This still works on CMake < 3.18 - by using older code, and if we detected CMake 3.18 or above - use CMake functions.

If we want to close https://github.com/SerenityOS/serenity/issues/9866 - we need to remove the checks in all modified functions, and upgrade the CMake version requested on the main CMakeLists.txt.

### CODE REVIEW

File: Meta/CMake/time_zone_data.cmake [hunk 1]
```diff
@@ -38,30 +38,15 @@ set(TZDB_SOUTH_AMERICA_PATH "${TZDB_PATH}/${TZDB_SOUTH_AMERICA_SOURCE}")
...
-    extract_tzdb_file("${TZDB_ANTARCTICA_SOURCE}" "${TZDB_ANTARCTICA_PATH}")
-    extract_tzdb_file("${TZDB_ASIA_SOURCE}" "${TZDB_ASIA_PATH}")
-    extract_tzdb_file("${TZDB_AUSTRALASIA_SOURCE}" "${TZDB_AUSTRALASIA_PATH}")
-    extract_tzdb_file("${TZDB_BACKWARD_SOURCE}" "${TZDB_BACKWARD_PATH}")
-    extract_tzdb_file("${TZDB_ETCETERA_SOURCE}" "${TZDB_ETCETERA_PATH}")
-    extract_tzdb_file("${TZDB_EUROPE_SOURCE}" "${TZDB_EUROPE_PATH}")
-    extract_tzdb_file("${TZDB_NORTH_AMERICA_SOURCE}" "${TZDB_NORTH_AMERICA_PATH}")
-    extract_tzdb_file("${TZDB_SOUTH_AMERICA_SOURCE}" "${TZDB_SOUTH_AMERICA_PATH}")
-    extract_tzdb_file("${TZDB_ZONE_1970_SOURCE}" "${TZDB_ZONE_1970_PATH}")
+    message(STATUS "Extractintg ${TZDB_ZIP_PATH} into ${TZDB_PATH}")
```
Comment (FireFox317) [inline Meta/CMake/time_zone_data.cmake:45]:
```suggestion
    message(STATUS "Extracting ${TZDB_ZIP_PATH} into ${TZDB_PATH}")
```
Same in the other file

Comment (elcuco) [reply Meta/CMake/time_zone_data.cmake]: Done, thanks! (and all the other hints you gave me in the discord channel!)

**Comment by bgianfo** [comment]
@ADKaster / @trflynn89 opinions here? Personally, it doesn't seem worth it to bump the required cmake just for this?

File: Meta/CMake/time_zone_data.cmake [hunk 2]
```diff
@@ -38,30 +38,15 @@ set(TZDB_SOUTH_AMERICA_PATH "${TZDB_PATH}/${TZDB_SOUTH_AMERICA_SOURCE}")
 set(TZDB_ZONE_1970_SOURCE zone1970.tab)
 set(TZDB_ZONE_1970_PATH "${TZDB_PATH}/${TZDB_ZONE_1970_SOURCE}")
 
-function(extract_tzdb_file source path)
-    if(EXISTS "${TZDB_ZIP_PATH}" AND NOT EXISTS "${path}")
-        message(STATUS "Extracting TZDB ${source} from ${TZDB_ZIP_PATH}...")
-        execute_process(COMMAND "${TAR_TOOL}" -C "${TZDB_PATH}" -xzf "${TZDB_ZIP_PATH}" "${source}" RESULT_VARIABLE tar_result)
```
Comment (ADKaster) [inline Meta/CMake/time_zone_data.cmake:44]: keep this function around, and and wrap the execute_process and version check in a version check for 3.18

we can use the pattern `file(ARCHIVE_EXTRACT INPUT "${zip_path}" DESTINATION "${dest_dir}" PATTERNS "${source_path}")

to do the same thing that the manual extraction is doing. same for the unicode file.

**Review by ADKaster — CHANGES_REQUESTED** [review]
This is not a required version bump, and there's a better way to use file(EXTRACT_ARCHIVE, as mentioned on discord. ZIP_LISTS and EXTRACT_ARCHIVE are nice, but not a killer enough feature to bump just yet.

Note that that this is related to https://github.com/SerenityOS/serenity/issues/9866 as well. Not sure if it completely closes it, but I don't think there's many more uses of those tools outsid ethese methods.

File: Meta/CMake/time_zone_data.cmake (hunk 2 above)
Comment (elcuco) [reply Meta/CMake/time_zone_data.cmake:44]: We have two options:

1. For unix targets, cmake 3.16 can be used (I will revert code), and for Windows - 3.18 will be needed.
2. Bump cmake requirement 3.18, and I will also restore the functionality to extract single files.

I propose a vote: thumbs up, means 1, down means 2. I will follow what ever you decide.

File: Meta/CMake/check_for_dependencies.cmake [hunk 3]
```diff
@@ -1,23 +0,0 @@
-#
-# Check for the dependencies that the Serenity (target) and Lagom (host) builds require.
-#
-
-# FIXME: With cmake 3.18, we can change unzip/untar/gzip steps to use
-#        file(  ARCHIVE_EXTRACT) instead
-#
-#        Additionally we have to emit an error message for each tool,
-#        as REQUIRED only works with cmake 3.18 and above.
-find_program(UNZIP_TOOL unzip REQUIRED)
```
[the remark this answers is not in the document]
Comment (trflynn89) [reply Meta/CMake/check_for_dependencies.cmake:10]: Also `GZIP_TOOL` is used elsewhere, e.g. unzipping usb.ids.gz

**Comment by elcuco** [comment]
New code: now I just modified the extraction functions, keeping the original idea: just extract needed files. On runtime - I check for CMake version, and if its 3.18 and above - use CMake directly. I am able to finish configuration on Windows... lets see on Linux/CI (I am optimistic!)

File: CMakeLists.txt [hunk 4]
```diff
@@ -240,42 +240,24 @@ endif()
...
-if(EXISTS ${PCI_IDS_GZ_PATH} AND NOT EXISTS ${CMAKE_STAGING_PREFIX}/${PCI_IDS_INSTALL_PATH})
-    message(STATUS "Extracting PCI ID database from ${PCI_IDS_GZ_PATH}...")
-    file(MAKE_DIRECTORY "${CMAKE_STAGING_PREFIX}/${CMAKE_INSTALL_DATAROOTDIR}")
-    if (CMAKE_VERSION VERSION_LESS 3.18.0)
-        execute_process(COMMAND "${GZIP_TOOL}" -d -c "${PCI_IDS_GZ_PATH}" OUTPUT_FILE "${CMAKE_STAGING_PREFIX}/${PCI_IDS_INSTALL_PATH}")
-    else()
-        file(ARCHIVE_EXTRACT INPUT "${PCI_IDS_GZ_PATH}" DESTINATION "${PCI_IDS_GZ_PATH}" PATTERNS "${CMAKE_STAGING_PREFIX}/${PCI_IDS_INSTALL_PATH}")
-    endif()
+if(ENABLE_PCI_IDS_DOWNLOAD AND NOT EXISTS ${CMAKE_STAGING_PREFIX}/${PCI_IDS_INSTALL_PATH} )
+    download_file("${PCI_IDS_URL}" "${CMAKE_STAGING_PREFIX}/${PCI_IDS_INSTALL_PATH}")
```
Comment (ADKaster) [inline CMakeLists.txt:253]: I'm not sure about the assertion in the commit message that downloading 2 MiB of data vs 500 KiB is "meaningless". On my internet connection, it doesn't matter, but there's contributors with much less powerful internet connections.

File: CMakeLists.txt [hunk 5]
```diff
@@ -253,7 +253,11 @@ endif()
 if(EXISTS ${PCI_IDS_GZ_PATH} AND NOT EXISTS ${CMAKE_STAGING_PREFIX}/${PCI_IDS_INSTALL_PATH})
     message(STATUS "Extracting PCI ID database from ${PCI_IDS_GZ_PATH}...")
     file(MAKE_DIRECTORY "${CMAKE_STAGING_PREFIX}/${CMAKE_INSTALL_DATAROOTDIR}")
-    execute_process(COMMAND "${GZIP_TOOL}" -d -c "${PCI_IDS_GZ_PATH}" OUTPUT_FILE "${CMAKE_STAGING_PREFIX}/${PCI_IDS_INSTALL_PATH}")
+    if (CMAKE_VERSION VERSION_LESS 3.18.0)
```
Comment (ADKaster) [inline CMakeLists.txt]: The commit message needs updated to reflect that we aren't bumping the minimum required yet.

File: CMakeLists.txt (hunk 4 above)
Comment (elcuco) [reply CMakeLists.txt:253]: The last update to serenityOS's git (`git pull`) gave me these numbers:
```
remote: Total 3080 (delta 2417), reused 2379 (delta 2350), pack-reused 529
Receiving objects: 100% (3080/3080), 856.94 KiB | 369.00 KiB/s, done.
```
Which means, that on "every week" we add about 500kb of git metadata. 

Then on build we download 44M from CLDR.

File: CMakeLists.txt (hunk 5 above)
Comment (elcuco) [reply CMakeLists.txt]: Done, see [eab454f](https://github.com/SerenityOS/serenity/pull/15040/commits/eab454f867f88b061e4d2d3ad9797cff7db037b5).

**Review by ADKaster — APPROVED** [review]
Fair point on the other files we download. This looks good to me  then.

**Pull request merged**